src/components/Tooltip/Tooltip.tsx

```diff
@@ -0,0 +1,229 @@
+const tooltipID = `tooltip-${Math.floor(Math.random() * 900000) + 100000}`
+
+function isElementInViewport(
+  el: HTMLElement,
+  win = window,
+  docEl = document.documentElement
+): boolean {
+  const rect = el.getBoundingClientRect()
+  return (
+    rect.top >= 0 &&
+    rect.left >= 0 &&
+    rect.bottom <= (win.innerHeight || docEl.clientHeight) &&
+    rect.right <= (win.innerWidth || docEl.clientWidth)
+  )
+}
+
+export const Tooltip = (
+  props: TooltipProps & JSX.IntrinsicElements['span']
+): React.ReactElement => {
+  const { label, position, children, ...spanProps } = props
+  const wrapperRef = useRef<HTMLElement>(null)
+  const triggerElementRef = useRef<HTMLElement>(null)
+  const tooltipBodyRef = useRef<HTMLElement>(null)
+  const tooltipTrigger = React.cloneElement(Children.only(children), {
+    ref: triggerElementRef,
+  })
+  const [isVisible, setVisible] = useState(false)
+
+  const tooltipClasses = classnames('usa-tooltip__body', {
+    'is-set': isVisible,
+    'usa-tooltip__body--top': position === 'top',
+    'usa-tooltip__body--bottom': position === 'bottom',
+    'usa-tooltip__body--right': position === 'right',
+    'usa-tooltip__body--left': position === 'left',
+    'is-visible': isVisible,
+  })
+
+  useEffect(() => {
+    if (
+      triggerElementRef.current &&
+      tooltipBodyRef.current &&
+      wrapperRef.current
+    ) {
+      const tooltipTrigger = triggerElementRef.current
+      const tooltipBody = tooltipBodyRef.current
+      const wrapper = wrapperRef.current
+      tooltipTrigger.setAttribute('aria-describedby', tooltipID)
+      tooltipBody.setAttribute('id', tooltipID)
+
+      // Calculate sizing and adjustments for positioning
+      const tooltipWidth = tooltipTrigger.offsetWidth
+      const tooltipHeight = tooltipTrigger.offsetHeight
+      const offsetForTopMargin = Number.parseInt(
+        window.getComputedStyle(tooltipTrigger).getPropertyValue('margin-top'),
+        10
+      )
+      const offsetForBottomMargin = Number.parseInt(
+        window
+          .getComputedStyle(tooltipTrigger)
+          .getPropertyValue('margin-bottom'),
+        10
+      )
+      const offsetForTopPadding = Number.parseInt(
+        window.getComputedStyle(wrapper).getPropertyValue('padding-top'),
+        10
+      )
+      const offsetForBottomPadding = Number.parseInt(
+        window.getComputedStyle(wrapper).getPropertyValue('padding-bottom'),
+        10
+      )
+      // issue dealing with null here
+      const offsetForTooltipBodyHeight = Number.parseInt(
+        window.getComputedStyle(tooltipBody).getPropertyValue('height'),
+        10
+      )
+      const leftOffset = tooltipTrigger.offsetLeft
+      const tooltipBodyWidth = tooltipBodyRef.current.offsetWidth
+      const adjustHorizontalCenter = tooltipWidth / 2 + leftOffset
+      const adjustToEdgeX = tooltipWidth + TRIANGLE_SIZE + SPACER
+      const adjustToEdgeY = tooltipHeight + TRIANGLE_SIZE + SPACER
+
+      /**
```

Comment:
I think it would be good to take advantage of React's declarative nature here, and set these attributes (style, classes) using `useState` and pass them to the element(s) inside of the markup instead of manipulating the HTML elements directly

Reply:
very much agree about avoiding `element.style` since we have JSX `className`.  It is a refactor though and the way the code is written seems fine to me for a first pass (since it just copies uswds). We could backlog an issue about refactoring this. @suzubara thoughts?

Reply:
yes, I'm okay with that!